docs/proposals/29-autoscaling-storage-volumes.md

```diff
@@ -0,0 +1,335 @@
+endpoint by `prometheus-seed`.
+
+`pvc-autoscaler` runs in active/passive replication mode, based on the standard leader election mechanism, supplied by
+the K8s controller runtime library. Only the active replica is shown in _Fig.1_.
+
+### PVC Autoscaler - Deployer Structure
+The `pvc-autoscaler` application and its supporting artefacts are deployed by `gardener`, as part of the seed
+reconciliation flow.
+
+The `vali` and `prometheus` deployers, part of `gardenlet`'s shoot reconciliation flow, ensure that the PVC template section in
+the respective StatefulSet objects contains the annotations, necessary to instruct `pvc-autoscaler` how to act on the
+PVCs, created from those templates.
+
+That `gardenlet` action mechanism, based on PVC templates, only affects PVCs upon creation. Therefore, it has no effect
+on existing shoots, and on the observability PVCs of existing seed clusters. To enable PVC autoscaling for existing
+shoots, those PVCs will be directly annotated by the operator. Such direct intervention allows gradual transition,
+where autoscaling can initially be enabled only for a few PVCs for validation purposes.
+Similarly, to enable PVC autoscaling on existing seed clusters' `garden` PVCs, those will be directly annotated at
+operators' discretion.
+For details regarding the transition of existing volumes to autoscaling, see 
+[section 'Enabling Autoscaling for Existing Volumes'](#enabling-autoscaling-for-existing-volumes)
+
+![02-operator_structure.png](assets/gep29-02-operator_structure.png)
+
+_Fig.2: Operator structure_
+
+VPA is used to scale `pvc-autoscaler`. For the future security enhancement scenario, where `kube-rbac-proxy` is used
+(see [section 'Future Enhancements'](#future-enhancements)),
+that container is expected to be excluded from autoscaling, and to run with fixed resource requests. Pod evictions
+driven by that lean container are unjustified, as they would only achieve minimal resource savings, and disrupt
+the main container.
+
+The feature is deployed behind a dedicated feature gate. When disabled, `pvc-autoscaler` is removed, and related
+annotations are removed from the shoot control plane observability StatefulSets, and the seed observability StatefulSets.
+
+#### Initial volume size and MaxAllowed
+
+| Component type | Component            | Old size | Initial size    | Max size |
+|----------------|----------------------|----------|-----------------|----------|
+| Shoot          | prometheus           | 20Gi     | 1Gi             | 40Gi     |
+| Shoot          | vali                 | 30Gi     | 1Gi             | 60Gi     |
+| Seed           | prometheus-aggregate | 20Gi     | 2Gi             | 40Gi     |
+| Seed           | prometheus-cache     | 10Gi     | 6Gi<sup>*</sup> | 20Gi     |
+| Seed           | prometheus-seed      | 100Gi    | 5Gi             | 200Gi    |
+| Seed           | vali                 | 100Gi    | 5Gi             | 200Gi    |
```

Comment:
Excellent point! I had based the proposal on the overall nature of the observability workloads, which to my knowledge use log-structured DBs, and thus should be expected to do mostly sequential I/O. Plus a measurement of the data volume load patterns.

Your question prompted me to do a more detailed analysis of real-life instances. It's not a trivial matter, but from what I see, we should be ok with the smaller volumes. Prometheus does a small number of relatively larger writes. Overall it is not too storage-intensive. Vali is the demanding case. It has noteworthy sustained load, and some very sharp spikes. With it, per-GB quota on GCP would not alone suffice to cover the sustained load, and the pod would rely on some of the node's baseline IOPS quota. But even with a node that's loaded only with Vali pods, the baseline quota should suffice, if my guesstimates are correct.

In the extreme case where a shoot is created and immediately loaded to generate extreme amount of logs, it could use a substantial fraction of the node's baseline quota (realistically, about 1/3) until its volumes fill up and trigger expansion (which would come up with its own per-GB quota). Even then I don't expect a practical problem, except in case of deliberate DoS attack, and even that I expect would result in just a bit of delay in log collection.

As to the extreme load spikes I see, those will definitely dip into node/burst quotas, but that's mostly what those quotas are for, anyway.